File: 00424.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # start position of the sliding window
        start = 0

        # most freq char till now (from index 0)
        # or you could think about it in this way:
        # the most freq char in the longest substring so far
        maxCount = 0

        # longest substring so far
        maxLength = 0

        # freq for the current sliding window
        count = {}

        # set the initial window to be as far as possible
        # and then looks for any substring longer than that
        for end in range(len(s)):
            count[s[end]] = count.get(s[end], 0) + 1
            maxCount = max(maxCount, count[s[end]])
            
            # the idea is:
            # the length of the substring should be <= most freq char + k
            if end - start + 1 > k + maxCount:
                count[s[start]] -= 1
                start += 1
                update maxCount
            else:
                # update the longest substring if it meets the requirement
                maxLength = max(maxLength, end - start + 1)
        return maxLength


def main():
    s = Solution()
    s1 = "ABAB"
    k = 2
    print(s.characterReplacement(s1, k))

    s1 = "AABABBA"
    k = 1
    print(s.characterReplacement(s1, k))

    s1 = "ABBB"
    k = 2
    print(s.characterReplacement(s1, k))

    s1 = "BAAAB"
    k = 2
    print(s.characterReplacement(s1, k))

    return




if __name__ == '__main__':
    main()

File: 00424.c
#define MAX(a, b) ((a) > (b) ? (a) : (b))

int getMaxCount (const int count[])
{
    int maxCount = 0;


    for (int ii = 'A'; ii <= 'Z'; ii++) {
        maxCount = MAX(maxCount, count[ii]);
    }

    return maxCount;
}

int characterReplacement (const char * s, const int k)
{
    int count['Z' + 1] = {0};
    int maxCount;
    int maxLen = 0;
    int start = 0;
    int end = 0;


    while (end < strlen(s)) {
        count[s[end]]++;
        maxCount = getMaxCount(count);

        if (end - start + 1 > k + maxCount) {
            // decrease the counter on [start]
            count[s[start]]--;
            start++;
        } else {
            maxLen = MAX(maxLen, end - start + 1);
        }
        // here is the tricky point:
        // no matter if the condition is met or not, the end variable gets increased
        // because we don't want to shrink the window
        // Think about it in this way: we first expand the window as long as possible
        // and then we have no interest on any smaller window even if it meets the condition
        end++;
    }

    return maxLen;
}
